### ISSUE TITLE
Lint for generalizeable captures

### ISSUE THREAD

**Issue by est31** [rust-lang/rust-clippy#10530]
### What it does

Take this code:
```Rust
fn foo(v: Option<char>) -> u32 {
    match v {
        Some(w @ 'a' | w @ 'c' | w @ 'd') => w as u32,
        Some(_) => 1,
        None => 0,
    }
}
```

The binding `w` of the first arm's pattern is repeated three times. One can simplify the pattern as follows:

```Rust
fn foo(v: Option<char>) -> u32 {
    match v {
        Some(w @ ('a' | 'c' | 'd')) => w as u32,
        Some(_) => 1,
        None => 0,
    }
}
```

### Lint Name

repetitive_capture

### Category

style

### Advantage

This saves repetition of the binding (less chars) and makes the code more maintainable as renames only affect one site instead of multiple

### Drawbacks

It adds one layer of `()`s, but this should be bearable

### Example

provided above

**Comment by est31** [comment]
See also https://github.com/rust-lang/rust/pull/109489 where I found instances of this in rustc.

**Comment by Almugra** [comment]
I think [unnested_or_patterns](https://rust-lang.github.io/rust-clippy/master/index.html#unnested_or_patterns) should already take care of that issue

**Comment by est31** [comment]
Hmm good point about unnested_or_patterns. I have tried enabling all the clippy lints before I filed the issue but it didn't show up, not sure what went wrong back then.

```
error: unnested or-patterns
 --> src/main.rs:5:9
  |
5 |         Some(w @ 'a' | w @ 'c' | w @ 'd') => w as u32,
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnested_or_patterns
note: the lint level is defined here
 --> src/main.rs:1:8
  |
1 | #[deny(clippy::pedantic)]
  |        ^^^^^^^^^^^^^^^^
  = note: `#[deny(clippy::unnested_or_patterns)]` implied by `#[deny(clippy::pedantic)]`
help: nest the patterns
  |
5 |         Some(w @ ('a' | 'c' | 'd')) => w as u32,
  |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Closing as it's already implemented in clippy.
